Data Security Plugin API (CORE-207,CORE-763,CORE-768)

rvesse opened the pull request:
This is an updated PR for PR #68, it brings the Security Plugin API draft branch up to date with current development and makes some other improvements:

- Simplifies the API to focus **only** on data security in light of our move to an OAuth 2/OIDC compliant authentication server that provides a standards compliant mechanism for obtaining user attributes
    - Various API renamed/removed to reflect this
- Intentionally removes code around obtaining user attributes the old way
- Intentionally makes AuthZ always on when AuthN is enabled
- 100% test coverage across the new Data Security Plugin API and RDF-ABAC plugin implementation

rvesse committed:
1st draft of Security Plugin API (CORE-207)

Outlines the basic interfaces and model classes necessary for the
Security Plugin proposal to help flesh out the design and identify areas
where the paper sketch doesn't match implementable reality.

Starts writing up documentation around it, from that work and additional
tests written in support of documentation make some implementation
changes to avoid generic signature issues when used by applications.

rvesse committed:
2nd Draft of Security Plugin API design (CORE-207)

Co-authored-by: Paul Gallagher <132362215+TelicentPaul@users.noreply.github.com>
Co-authored-by: Rob Walpole <rob.walpole@telicent.io>
Co-authored-by: Rob Levi <rob.levi@telicent.io>

- Add new RequestContext API
- JAX-RS implementation of RequestContext API
- Adds support for the new Security Plugin RequestContext into the JAX-RS
  Base Server module including some basic test coverage of its
  functionality
- Retrieving Entitlements is now based on passing in a RequestContext so
  the plugin has more information to work with.
- Authorizer now has separate canRead(), canWrite() and canUse() methods
  for making different kinds of access decisions.

rvesse committed:
Add sequence diagrams for label flows (CORE-207)

rvesse committed:
Add logging warning (CORE-207)

rvesse committed:
3rd draft of Security Plugin API

- Change Entitlements terminology back to User Attributes
- Add class diagram
- Clarifications around role of Authorizer as a PDP and/or PEP
- Clarifications around AttributesProvider as PCP
- Outlined key risks explicitly in the design document

rvesse committed:
Clean up bad rebase induced merge conflicts (CORE-207, CORE-763)

Restores correct dependencies into various modules to get a clean build

rvesse committed:
Use latest RDF ABAC libraries (CORE-207,CORE-768)

This allows to avoid some unnecessary conversion from String to byte[]
and just copy the byte[] sequences returned from the LabelsStore
directly

rvesse committed:
Fix test instabilities and clean up (CORE-763,CORE-768)

Fixes up issues in jaxrs-base-server module tests where new tests around
Security Plugin API uncovered a test harness isolation issue with
MockKeyServer.

Also made some further improvements to security-plugin-rdf-abac module
based on using newer APIs from RDF ABAC 1.0.0

robwtelicent commented on docs/security/index.md:
```diff
@@ -0,0 +1,722 @@
+`prepareAuthorizer()` method passing in the user attributes.  `Authorizer` instances are intended to be scoped to the
+lifetime of a single user request so they may cache any access decisions if they encounter the same labels repeatedly.
+
+Once an application has an `Authorizer` it calls one of the `canRead()`/`canWrite()`/`canUse()` methods passing in
```
So `canUse()` isn't about the data it is about whether you can even ask about the data, whereas `canRead()` and `canWrite()` are about the data?

rvesse replied:
Yeah `canUse()` is not necessarily the best naming, perhaps `canMakeRequest()` would be clearer?

As it could be applied to any API request, even those that have nothing to do with data access

robwtelicent replied:
Yes, that would be clearer - or perhaps just `isAuthorized()`?

rvesse committed:
Minor typo fix (CORE-207)

Co-authored-by: Rob Walpole <rob.walpole@telicent.io>

rvesse committed:
Rename canUse() to canMakeRequest() (CORE-207)

This clarifies the purpose of this authorization decision method and its
intended usage.

Co-authored-by: Rob Walpole <rob.walpole@telicent.io>

TelicentPaul reviewed (approved):
My comments below are a little disjointed. Probably more to follow as this will obviously be an ongoing conversation.

Right now my thinking is that we should force the plugin resolution at start-up. The Failsafe Plugin is a good idea but misconfiguration should be identified as soon as possible especially for something so important. The potential gap between a "successful" deployment, a user discovering a problem, and then the error being reported too wide to be comfortable. 

I can't envisage plug-ins being too complicated but we may want to think about a health-check mechanism. Not for now, as it would just be clutter in the interface, but something for the future - imagine a plugin that queries a DB, what if that's down? 
Do we want plug-in info (version, type, what-have-you) reportable via the server-info endpoint or it's own endpoint?

Similarly, we'll want to separate the plug-ins to their own repo(s) akin to the doc pipeline. For now, best to keep things simple. For demonstrative purposes, we may wish to knock together another plug-in even if stupidly simple.

Metrics are something else to consider - do we want to track how many parsing failures, malformed labels, malformed attributes etc...? I suppose we could wrap a plug-in with another like chinese dolls.  Might help for debugging I suppose = )


Note: I've carried out a rebase so that it's in-line with the current main.

rvesse commented:
Some of the ideas here are being superseded by @tatelicent's reworked authentication approach as seen in PR #196 so at some point once that's all fully deployed need to rework this PR to remove the API authorisation aspects as that's being handled by a different approach than was originally envisaged in this PR

rvesse committed:
Simplify Plugin API (CORE-207,CORE-763)

- Remove unnecessary Authorizer methods since those use cases are now
  covered via new Auth server
- Use new AttributesStoreAuthServer where appropriate
- Add label parsing and evaluation caches into RDF-ABAC plugin
  information
- Make cache sizes and durations configurable where appropriate
- Avoid unnecessary re-encoding of AttributeValueSet
- Support only parsing new style attributes

rvesse committed:
Improved RDF-ABAC plugin test coverage (CORE-207,CORE-763)

Improves the RDF-ABAC security plugin test coverage to ensure high
coverage across key implementation classes.

rvesse committed:
Simplify API (CORE-207,CORE-763)

Remove concept of Schema prefix since we have no intention of ever
supporting multiple labelling schemas in a single installation.

Also improves RDF-ABAC plugin implementation and tests to reach 100%
test coverage while eliminating defunct code

rvesse committed:
Further improve test coverage (CORE-207,CORE-763)

Further improves the unit test coverage around the new Security Plugin
APIs achieving 100% coverage on the current Plugin API and
implementations.

rvesse committed:
Remove defunct IdentityProvider concept (CORE-207)

rvesse committed:
Rename modules and APIs for clarify (CORE-207,CORE-763)

Renames the new data security plugin API modules and key classes to make
it explicit that these pertain to data security and not general
application security which is handled in other ways

rvesse committed:
Updated documentation (CORE-207,CORE-763)

Renames and updates documentation around data security plugins to make
their purpose and usage clearer

rvesse committed:
Documentation tidy up (CORE-207,CORE-763)

Bring documentation up to date with Data Security Plugin API current
status, also tidy up JAX-RS documentation to remove mention of
deprecated/removed functionality.

rvesse committed:
Code and documentation tidy up (CORE-207,CORE-763)

- Corrects various JavaDoc typos or outdated references
- Renames Authorizer to DataAccessAuthorizer for clarity
- Further tidies up and improves documentation
- Fix license headers

TelicentPaul reviewed (commented):
Looks good.  It reduces the coupling and gives a clear approach. 

First draft and all that but would we ever consider a plug-in per dataset as opposed to the global singleton? 

Might be worth for illustrative purposes implementing say a cached rdf-abac-plugin to show the variety/flexibility of the plugins, or something else. A no-op plugin comes to mind but that's just a bit lazy on my end. 

For the Request Context, JWT & UserInfo suits our purposes now but I keep thinking of the issues with the Auth Server & the Service Accounts where the user model doesn't quite fit. Probably a stretch here but shoehorning something like that in later, wouldn't be ideal.

robwtelicent committed:
Cleaning up following rebase

robwtelicent committed:
Updating security plugin for latest LabelStore and adding more tests

robwtelicent committed:
Adding  label backup/restore and decideDataset functionality

robwtelicent committed:
Adding label compact funtionality

rvesse commented on data-security-parent/data-security-plugin-rdf-abac/src/main/java/io/telicent/smart/cache/security/data/plugins/rdf/abac/RdfAbacPlugin.java:
```diff
@@ -0,0 +1,228 @@
+        if (context == null) {
+            return FailSafeAuthorizer.INSTANCE;
+        } else {
+            // TODO Compute attributes from UserInfo here
```
Outdated TODO comment, this is implemented

robwtelicent replied:
Removed.

rvesse commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/plugins/DataSecurityPlugin.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @return Labels backup
+     */
+    Optional<SecurityLabelsBackup> prepareLabelsBackup();
```
Is it worth giving all these methods that return `Optional<?>` a default implementation that returns `Optional.empty()`?

robwtelicent replied:
Makes sense! Updated.

rvesse commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/labels/SecurityLabelsRestore.java:
```diff
@@ -0,0 +1,34 @@
+     * @param restorePath  the path from which to restore the security labels
+     * @param node         the JSON object node containing backup metadata
+     */
+    void restore(DatasetGraph datasetGraph, String restorePath, ObjectNode node);
```
One particular quirk here is that this bleeds implementation detail about how Backup/Restore requests are handled on the SC-Graph side into the generic portion of the API

I wonder whether we ought to migrate some of the generic Backup/Restore APIs that @karolina-telicent wrote for Smart Cache Storage into Smart Caches Core to avoid this?

e.g. `RestoreConfig` and `RestoreStatus`

Then you can have `RdfAbacLabelsRestore` simply take the `ObjectNode` in as a configuration parameter in the generic API and restrict the relevant JSON dependencies to the RDF ABAC plugin

robwtelicent replied:
Created [CORE-1365](https://telicent.atlassian.net/browse/CORE-1365) to address this.

rvesse commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/labels/SecurityLabelsBackup.java:
```diff
@@ -0,0 +1,34 @@
+     * @param backupPath the path to back up the labels to
+     * @param node the result node
+     */
+    void backup(DatasetGraph datasetGraph, String backupPath, ObjectNode node);
```
Similar concern as with `SecurityLabelsRestore`

robwtelicent replied:
Created [CORE-1365](https://telicent.atlassian.net/browse/CORE-1365) to address this.

rvesse commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/plugins/DataSecurityPluginLoader.java:
```diff
@@ -0,0 +1,153 @@
+     * <p>
+     * This means that the next call to {@link DataSecurityPluginLoader#load()} will reload the plugin.  This is
+     * primarily only of use in unit and integration test scenarios where you need to change the configuration
+     * information that the
```
Unfinished Javadoc, thanks past me 🤣

robwtelicent replied:
I asked Claude to finish the sentence (as working out the most probable next word is supposed to be what LLMs are good at!). See if you agree..

rvesse commented on jaxrs-base-server/src/main/java/io/telicent/smart/cache/server/jaxrs/applications/AbstractApplication.java:
```diff
@@ -60,18 +58,10 @@ public Set<Class<?>> getClasses() {
-                        "The Authorization Policy enforcement feature has been explicitly disabled via configuration variable {}.  Any authorization policy defined on application resources will be ignored and NOT enforced.  If this was not intended please ensure that this feature flag is set to true in your environment.",
-                        AuthConstants.FEATURE_FLAG_AUTHORIZATION);
-            }
+            classes.add(JwtAuthFilter.class); // Enforces Authentication (valid, signed JWT presented)
```
Makes sense, we always intended for the Authorisation feature flag to be temporary so as part of this change is as good a time as any to remove it

(Although now I've written this I'm wondering if I made this part of the commit...)

robwtelicent replied:
Hmm. It looks like this got removed when I was cleaning up after my initial rebase and lots of things were broken. I'm not sure now it was intentional on my part or more of a best-guess at where it should be - but if you're happy with it then I'll leave it as it is!

rvesse commented on docs/data-security/design.md:
Probably need some fairly extensive updates to this to align with current implementation

robwtelicent replied:
Updated.

rvesse commented on docs/data-security/index.md:
```diff
@@ -0,0 +1,73 @@
+`SecurityLabels<?>` and pass those to `DataAccessAuthorizer.canRead(SecurityLabels<?>)` to test whether labelled data
+should be visible to the user.
+4. `close()` the authorizer to free any resources and request state it is holding.
+
```
Probably wants some example of how the new APIs allow some of the other functionalities that are being migrated from RDF-ABAC/SC-Graph to be utilised

robwtelicent replied:
Updated.

rvesse commented on jaxrs-base-server/src/test/resources/logback-test.xml:
```diff
@@ -35,7 +35,7 @@
     <logger name="org.glassfish.jersey.internal.Errors" level="ERROR" />
 
-    <root level="OFF">
+    <root level="INFO">
```
```suggestion
    <root level="OFF">
```

Revert to `OFF` once we're happy with the rest of this PR and don't need higher logging level to debug any impacts on JAX-RS Server tests

robwtelicent commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/labels/SecurityLabelsApplicator.java:
```diff
@@ -0,0 +1,44 @@
+     * @param triple Triple
+     * @return Security Label, or {@code null} if no security label applies
+     */
+    SecurityLabels<?> labelForTriple(Triple triple);
```
A `labelForQuad(Quad quad)` method has been added in [this commit](https://github.com/telicent-oss/smart-caches-core/pull/150/changes/25382ba5bad6048bec31a5e9c390dfaf46af4d6d).

TelicentPaul commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/plugins/DataSecurityPlugin.java:
```diff
@@ -0,0 +1,179 @@
+     * </p>
+     * <p>
+     * In the event that a {@code null} context is provided, or there's insufficient user information to make
+     * authorization decisions then a plugin <strong>MUST</strong> return an authorizer that rejects all authorization
```
This doesn't match up with the RDF ABAC Plugin behavior below

TelicentPaul commented on data-security-parent/data-security-plugin-rdf-abac/src/main/java/io/telicent/smart/cache/security/data/plugins/rdf/abac/RdfAbacPlugin.java:
```diff
@@ -0,0 +1,227 @@
+            if (context.userInfo() != null) {
+                attributes = toAttributeValueSet(context.userInfo().getAttributes());
+            } else {
+                attributes = AttributeValueSet.EMPTY;
```
If you look at the commentary on the method in the Data Security Plugin interface above, if we don't have any User Info data we should return FailSafeAuthorizer.INSTANCE so that everything is rejected

robwtelicent replied:
Agree. Updated.

TelicentPaul commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/requests/RequestContext.java:
```diff
@@ -0,0 +1,86 @@
+     * </p>
+     *
+     * @param header Header name
+     * @return Values, an empty list if no such headers are present
```
This doesn't match the behaviour in Jax RS Request Context below

TelicentPaul commented on jaxrs-base-server/src/main/java/io/telicent/smart/cache/server/jaxrs/auth/JaxRsRequestContext.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    public List<String> requestHeader(String header) {
+        return this.request.getHeaders().get(header);
```
If the header doesn't exist this will return null. The comments on the Reqeust Context interface says it'll be an empty list.  Given the underlying headers are a Multi value map (i.e. a Map of Lists) we can get away with this

```suggestion
        return this.request.getHeaders().getOrDefault(header, Collections.emptyList());
```

robwtelicent replied:
Updated.

TelicentPaul commented on data-security-parent/data-security-plugin-rdf-abac/pom.xml:
```diff
@@ -0,0 +1,127 @@
+        <dependency>
+            <groupId>org.apache.jena</groupId>
+            <artifactId>jena-fuseki-main</artifactId>
+            <version>6.0.0</version>
```
Utterly pedantic - we don't want to have fixed versions locked here

robwtelicent replied:
Agree. Updated.

TelicentPaul commented on data-security-parent/data-security-plugin-rdf-abac/src/main/java/io/telicent/smart/cache/security/data/plugins/rdf/abac/RdfAbacChangesApplyWithLabels.java:
```diff
@@ -0,0 +1,136 @@
+        } else {
+            if (this.targetGraph != null) {
+                g = targetGraph;
+            }
```
```suggestion
            } else if (g == null) {
                g = Quad.defaultGraphIRI;
            }
```
To mirror the behaviour of the add() method above.

robwtelicent replied:
Updated.

TelicentPaul commented on data-security-parent/data-security-core/src/main/java/io/telicent/smart/cache/security/data/plugins/DataSecurityPluginLoader.java:
```diff
@@ -0,0 +1,153 @@
+            if (PLUGIN != null) {
+                LOGGER.warn(
+                        "Resetting Telicent Security Plugin, if you see this message in a production environment please contact an Administrator immediately");
+            }
```
Pedantic of me. 
We'd potentially leak resources with future implementations here. 
Caveat - I know we're not really using it but for testing and close()'s a no-op in the given implementations.... but.... doesn't hurt.

```suggestion
           PLUGIN.close();
            }
```

robwtelicent replied:
Added.
